### PR TITLE
feat: ability to read macros from manifest

### PR DESCRIPTION
**Pull request by arteevraina**
This pull request add the ability to parse and read the `macros` array defined in the manifest file for a preprocessor under `preprocess` section.

### CODE REVIEW

**Comment by arteevraina** [comment]
Hi @milancurcic , Should I add a new example package to test this feature or should I populate the macros array in the example `preprocess_cpp` ?

**Comment by milancurcic** [comment]
Good question, I'd add it to the existing example package so that one example package describes one complete feature (preprocessing) rather than having a package for each sub-feature.

**Comment by arteevraina** [comment]
> Good question, I'd add it to the existing example package so that one example package describes one complete feature (preprocessing) rather than having a package for each sub-feature.

Sure, added it.

**Comment by awvwgk** [comment]
What about macros which access information from the manifest like the version numbers?

File: src/fpm_compiler.f90 [hunk 1]
```diff
@@ -416,12 +416,79 @@ subroutine set_preprocessor_flags (id, flags)
     do i = 1, size(package%preprocess)
         if (package%preprocess(i)%name == "cpp") then
             flags = flag_cpp_preprocessor// flags
+            call get_macros_from_manifest(id, flags, i)
             exit
         end if
```
Comment (milancurcic) [inline src/fpm_compiler.f90:421]: Should we add:

```fortran
else
    write(stderr, '(a)') 'Warning: preprocessor ' // package%preprocess(i)%name // ' is not supported; will ignore it'
```
or similar? Otherwise the user gets a silent ignore.

More generally, @fortran-lang/fpm, does fpm have a style recommendation on emitting warnings or errors from fpm's internal modules? Is it OK to do it and is there a preferred format?

File: src/fpm_compiler.f90 [hunk 2]
```diff
@@ -416,12 +416,79 @@ subroutine set_preprocessor_flags (id, flags)
...
     end do
 
 end subroutine set_preprocessor_flags
 
+!> This subroutine will check and get all the macros that are defined in the manifest file
+!> under preprocess section for a specific preprocessor
+subroutine get_macros_from_manifest(id, flags, index_of_preprocessor)
```
Comment (milancurcic) [inline src/fpm_compiler.f90:428]: Is it possible to test this subroutine from fpm's test suite?

**Review by milancurcic — APPROVED** [review]

File: src/fpm_compiler.f90 (hunk 1 above)
Comment (awvwgk) [reply src/fpm_compiler.f90:421]: Not yet, I'm hoping to upgrade the manifest reading to the upcoming TOML Fortran 0.3.0 release.

This would allow us to create warnings and errors with rich context information as shown in https://toml-f.readthedocs.io/en/latest/tutorial/fpm-lint.html and https://toml-f.readthedocs.io/en/latest/how-to/error.html.

**Comment by milancurcic** [comment]
Does anybody understand why the docs build failed?

```
> Run ford docs.md
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.5/x64/bin/ford", line 8, in <module>
    sys.exit(run())
  File "/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/ford/__init__.py", line 633, in run
    proj_data, proj_docs, md = initialize()
  File "/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/ford/__init__.py", line 236, in initialize
    return parse_arguments(vars(args), proj_docs, directory)
  File "/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/ford/__init__.py", line 406, in parse_arguments
    md = markdown.Markdown(
  File "/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/markdown/core.py", line 96, in __init__
    self.registerExtensions(extensions=kwargs.get('extensions', []),
  File "/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/markdown/core.py", line 125, in registerExtensions
    ext.extendMarkdown(self)
TypeError: MarkdownInclude.extendMarkdown() missing 1 required positional argument: 'md_globals'
Error: Process completed with exit code 1.
```

At a glance it seems like a dependency API may have changed. I don't know why that would happen if the dependencies list is frozen (I assume it is).

**Comment by awvwgk** [comment]
> Does anybody understand why the docs build failed? I re-ran the CI jobs to check whether this is transient and it's not.

See https://github.com/Fortran-FOSS-Programmers/ford/issues/437

> I don't know why that would happen if the dependencies list is frozen (I assume it is).

The dependencies for the docs are not frozen, we update as ford and its dependencies are updating.

**Review by zoziha — COMMENTED** [review]
Thanks for sharing, some feedback is given here.

_PS. I was thinking, it seems that the compiler preprocessing flag is only affected by the top project `fpm.toml`, is the impact of dependencies not considered yet?_

File: src/fpm_compiler.f90 [hunk 3]
```diff
@@ -416,12 +416,79 @@ subroutine set_preprocessor_flags (id, flags)
...
+    type(error_t), allocatable :: error
+    character(len=:), allocatable :: flags
+    character(len=:), allocatable :: macro_definition_symbol
+    character(:), allocatable :: valued_macros(:)
+    character(len=:), allocatable :: version
+
+    integer :: i
+    integer :: index_of_preprocessor
+
+    call get_package_data(package, "fpm.toml", error)
```
[the remark this answers is not in the document]
Comment (zoziha) [reply src/fpm_compiler.f90:440]: At the same time, the `package` parameter in `set_preprocessor_flags` can also be passed in from `fpm.f90::build_model::Line81`.
(This may avoid opening and closing `fpm.toml` multiple times)

**Comment by ZedThree** [comment]
The issue with the docs should be fixed: this was a problem in our dependencies that is now fixed. The CI _should_ work fine when it next builds. Local checkouts might need to `pip install --update --force ford` until the next version is release (imminently).

File: src/fpm_compiler.f90 (hunk 1 above)
Comment (arteevraina) [reply src/fpm_compiler.f90:417]: > Should we add:
> 
> ```fortran
> else
>     write(stderr, '(a)') 'Warning: preprocessor ' // package%preprocess(i)%name // ' is not supported; will ignore it'
> ```
> 
> or similar? Otherwise the user gets a silent ignore.
> 

Thanks for noticing this. I fixed this in my new commits.

File: src/fpm_compiler.f90 (hunk 2 above)
Comment (arteevraina) [reply src/fpm_compiler.f90]: Added a test for this as well. I had to do a few minor changes in the subroutine to handle the edge cases that were causing the error.

**Comment by arteevraina** [comment]
> Thanks for sharing, some feedback is given here.
> 
> _PS. I was thinking, it seems that the compiler preprocessing flag is only affected by the top project `fpm.toml`, is the impact of dependencies not considered yet?_

Thanks for reviewing the PR. 
Yes, you are correct. The preprocessing flag is only affected by the top-level project `fpm.toml`. The impact of dependencies was not really discussed before in the meeting while planning for this feature. But, it looks like a reasonable point for discussion that I will discuss in today's GSoC Meeting.

**Comment by arteevraina** [comment]
Hi, @milancurcic @awvwgk . I have added the requested changes in this PR. Please let me know if any further changes are requested in the PR. 

Thanks :)

**Comment by milancurcic** [comment]
@awvwgk can this be merged? It looks good to me. Thanks, @arteevraina!

**Comment by arteevraina** [comment]
> @awvwgk can this be merged? It looks good to me. Thanks, @arteevraina!

Hi Milan, Just a few more changes left to be done that Sebastian and I discussed in the last meeting. I am currently working on them and then they can be merged.

**Review by awvwgk — CHANGES_REQUESTED** [review]
As discussed in the last meeting, the macro definitions are currently global rather than package local, which can lead to undesirable outcomes when building a large project with many dependencies.

**Comment by arteevraina** [comment]
Hi @milancurcic @awvwgk , I have made the changes to make macro definitions local to the package. Thanks :pray:

**Comment by awvwgk** [comment]
Can you add a package example where this behavior is shown?

**Comment by arteevraina** [comment]
@awvwgk I have added the example package to showcase this behaviour.

File: example_packages/preprocess_hello_dependency/src/preprocess_hello_dependency.f90 [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  implicit none
+  private
+
+  public :: say_hello
+contains
+  subroutine say_hello
+
+!> This build should break because we do not have FOO defined as macro in the fpm.toml file of
+!> this package.
+#ifndef FOO
```
Comment (awvwgk) [inline example_packages/preprocess_hello_dependency/src/preprocess_hello_dependency.f90:11]: If you use `#ifdef FOO` the package would pass the build and we can safely run it as check in the CI.

Comment (arteevraina) [reply example_packages/preprocess_hello_dependency/src/preprocess_hello_dependency.f90]: Yes, added the requested the changes.

**Review by awvwgk — APPROVED** [review]

**Review by zoziha — APPROVED** [review]
Thanks @arteevraina , LGTM ❤.

**Comment by arteevraina** [comment]
Hi @milancurcic, @awvwgk . Can we merge this one if no more changes are required in this ?

**Comment by milancurcic** [comment]
Great, thanks all!
